packages/framework/aqueduct/src/data-objects/migrationDataObject.ts

```diff
@@ -3,100 +3,130 @@
  */
 
-import { assert } from "@fluidframework/core-utils/internal";
-import type { IChannel } from "@fluidframework/datastore-definitions/internal";
-import { SharedDirectory, type ISharedDirectory } from "@fluidframework/map/internal";
-import type { ISharedObject } from "@fluidframework/shared-object-base/internal";
-import { SharedTree, type ITree } from "@fluidframework/tree/internal";
+import type {
+	IFluidDataStoreRuntime,
+	IChannelFactory,
+} from "@fluidframework/datastore-definitions/internal";
 
 import type { IDelayLoadChannelFactory } from "../channel-factories/index.js";
 
-import { dataObjectRootDirectoryId } from "./dataObject.js";
 import { PureDataObject } from "./pureDataObject.js";
-import { treeChannelId } from "./treeDataObject.js";
 import type { DataObjectTypes } from "./types.js";
 
 /**
- * ! TODO
+ * Descriptor for a model shape (arbitrary schema) the migration data object can probe for
+ * or create when initializing. The probe function may inspect multiple channels or other
+ * runtime state to determine whether the model exists and return a model instance.
+ * @legacy
+ * @alpha
+ */
+export interface ModelDescriptor<TModel = unknown> {
+	// Discriminated union tag
+	type?: string;
+	// Optional convenience id for simple channel-backed models
+	id?: string;
+	//* Consider if we want something more formal here or if "duck typing" the runtime channel structure is sufficient.
+	//* See Craig's DDS shim branch for an example of tagging migrations
+	// Probe runtime for an existing model based on which channels exist. Return the model instance or undefined if not found.
+	probe: (runtime: IFluidDataStoreRuntime) => Promise<TModel | undefined>;
+	/**
+	 * Load any delay-loaded factories needed for this model.
+	 *
+	 * @remarks
+	 * This must be called before create can be called - otherwise the factory may be missing!
+	 */
+	ensureFactoriesLoaded: () => Promise<void>;
+	/**
+	 * Synchronously create the model.
+	 * @remarks
+	 * Any delay-loaded factories must already have been loaded via ModelDescriptor.loadFactories.
+	 */
+	create: (runtime: IFluidDataStoreRuntime) => TModel;
+	/**
+	 * The factories needed for this Data Model, divided by whether they are always loaded or delay-loaded
+	 */
+	sharedObjects: {
+		//* Do we need to split these apart or just have IChannelFactory[]?
+		alwaysLoaded?: IChannelFactory[];
+		delayLoaded?: IDelayLoadChannelFactory[];
+	};
+	// Optional runtime type guard to help callers narrow model types.
+	//* Probably remove?  Copilot added it
+	is?: (m: unknown) => m is TModel;
+}
+
+/**
+ * This base class provides an abstraction between a Data Object's internal data access API
+ * and the underlying Fluid data model.  Any number of data models may be supported, for
+ * perma-back-compat scenarios where the component needs to be ready to load any version
+ * from data at rest.
  * @experimental
  * @legacy
  * @alpha
  */
 export abstract class MigrationDataObject<
```

Comment:
Is this flexible enough to also model the normal `DataObject`, and the current `TreeDataObject`, with the right descriptors, without needing to change persisted state? That would be very cool, because it would allow us switch Pages Components to this data object, even if we don't want to change its persisted state.